doc: Add documentation about Pin control (#7559)
nrfconnect/sdk-nrf: doc/nrf/ug_pinctrl.rst

.. _ug_pinctrl:

Pin control
###########

.. contents::
   :local:
   :depth: 2

Pin control, also referred to as ``pinctrl``, is a Zephyr RTOS API introduced as an effort to standardize how SoC peripheral pins are configured.
This enables mapping peripheral signals to the required pins as well as configuring necessary pin properties.
For example, using pin control, you can route ``UART0`` ``RX`` signal to pin ``P1.01`` and configure the pin pull-up resistor.

> Before pin control, the same functionality was available in a vendor specific way.
> For the previous example, before pin control, Nordic SoCs had to be configured in the following way:

Reviewer: ```suggestion
Before pin control, the same functionality was available in a vendor specific way.
For the previous example, Nordic SoCs had to be configured in the following way:
```

(repeats)

Author: Fixed

.. code-block:: dts

    &uart0 {
        ...
        rx-pin = <33>;
        rx-pull-up;
        ...
    };

> Pin control involves *states*, where each state encodes the pin configurations for a state of the peripheral.

Reviewer: ```suggestion
Pin control works using a **state model**, where each state encodes the pin configurations for a state of the peripheral.
```

note: `state` repeats too much?

Author: Fixed

Following are the standard states:

* ``default`` - Peripheral in working mode.

> * ``sleep`` - Peripheral in sleep mode.

Reviewer: ```suggestion
* ``sleep`` - Peripheral in sleep or low power mode.
```

A similar approach is also used in other systems like Linux.
For more details including nRF specific bits, see the next section.

> For more general information, see the :ref:`Zephyr documentation on pin controllers <pinctrl-guide>`.

Reviewer: ```suggestion
For more general information, see the :ref:`Zephyr documentation on pin control <pinctrl-guide>`.
```

Pin control in nRF
******************

The nRF pin controller is represented in :ref:`devicetree <dt-guide>` with the ``pinctrl`` node label.
This node contains all pin configurations and you can modify it as shown in the following snippet:

.. code-block:: dts

   &pinctrl {
           /* Add your modifications here */
   };

All device pin configurations must be placed in child nodes of the ``pinctrl`` node, as shown in the following example:

.. code-block:: dts

   &pinctrl {

>      /* configuration for uart0 device, default state */

Reviewer: ```suggestion
      /* configuration for uart0 device, default state */
```

* ``nordic,invert`` - Invert pin polarity  (set active state to low). This property is valid only for PWM channel output pins.

To link this pin configuration with a device, use a pinctrl-N property where N is the state index starting from zero.
You can add the following code in the DTS file of your board:

.. code-block:: dts

   #include "board-pinctrl.dtsi"

    &uart0 {
        ...
        /* default state */
        pinctrl-0 = <&uart0_default>;
        /* sleep state (only applicable if CONFIG_PM_DEVICE=y) */
        pinctrl-1 = <&uart0_sleep>;
        /* state assigned to each pinctrl-N property by index */
        pinctrl-names = "default", "sleep";
        ...
    };

Migration of the devicetree files
*********************************

To migrate the devicetree files of any nRF-based board, you must complete the following steps:

1. Create a :file:`board-pinctrl.dtsi` file, defining each peripheral state pin configuration.
2. Include this file in the :file:`board.dts` file.

> 3. Replace all ``-pin`` related properties with pin control states.

Reviewer: ```suggestion
3. Replace all ``-pin`` related properties with pin control states.
4. Enable pin control by adding ``CONFIG_PINCTRL=y`` to the :file:`board.conf`
```

The following example illustrates the migration of a board file:

.. code-block:: dts

    /* old board.dts */
    &uart0 {
        ...
        tx-pin = <5>;
        rx-pin = <33>;
        rx-pull-up;
        ...

>     };
> .. code-block:: dts

Reviewer: ```suggestion
    };

.. code-block:: dts
```

    /* new board-pinctrl.dtsi file */
    &pinctrl {
        uart0_default: uart0_default {
            group1 {
                psels = <NRF_PSEL(UART_TX, 0, 5)>;
            };
            group2 {
                psels = <NRF_PSEL(UART_RX, 1, 1)>;
                bias-pull-up;
            };
        };
        uart0_sleep: uart0_sleep {
            group1 {
                psels = <NRF_PSEL(UART_TX, 0, 5)>,
                        <NRF_PSEL(UART_RX, 1, 1)>;
                low-power-enable;
            };
        };

>     };
> .. code-block:: dts

Reviewer: ```suggestion
    };

.. code-block:: dts
```
